os/pm/pm_timer.c

```diff
@@ -0,0 +1,149 @@
+/****************************************************************************
+ *
+ * Copyright 2024 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/pm/pm.h>
+#include <tinyara/wdog.h>
+#include <tinyara/sched.h>
+#include <errno.h>
+#include "pm.h"
+
+/************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************/
+
+/************************************************************************
+ * Private Type Declarations
+ ************************************************************************/
+
+/************************************************************************
+ * Private Variables
+ ************************************************************************/
+
+/************************************************************************
+ * Private Functions
+ ************************************************************************/
+
+#if defined(CONFIG_PM_LCD_NORMAL_TO_IDLE_THRESH) && defined(CONFIG_PM_LCD_IDLE_TO_STANDBY_THRESH) && defined(CONFIG_PM_STANDBY_TO_SLEEP_THRESH)
+static uint32_t pm_gettimeinterval(void)
+{
+	uint32_t ret = 0;
+	if (g_pmglobals.state == PM_NORMAL) {
+		ret = CONFIG_PM_LCD_NORMAL_TO_IDLE_THRESH;
+	} else if (g_pmglobals.state == PM_IDLE) {
+		ret = CONFIG_PM_LCD_IDLE_TO_STANDBY_THRESH;
+	} else if (g_pmglobals.state == PM_STANDBY) {
+		ret = CONFIG_PM_STANDBY_TO_SLEEP_THRESH;
+	}
+	DEBUGASSERT(ret != 0);
+	return ret;
+}
+#endif
+
+static void pm_routine(int argc)
+{
+
+	/* After timer expiration change PM State to next low power state */
+	enum pm_state_e nextstate = pm_getnextstate();
+	pm_changestate(nextstate);
```

Comment:
if we execute pm_changestate in the ISR, we need to remove or seperate pm_register excuteing

It is big operation